identity_diff/src/vec.rs

```diff
@@ -0,0 +1,281 @@
+        let max = usize::max(l_len, r_len);
+        let mut changes: Vec<InnerVec<T>> = vec![];
+
+        for index in 0..max {
+            match (self.get(index), other.get(index)) {
+                (None, None) => panic!("No data to match"),
+                (Some(x), Some(y)) if x == y => {}
+                (Some(x), Some(y)) => changes.push(InnerVec::Change { index, item: x.diff(y) }),
+                (None, Some(x)) => changes.push(InnerVec::Add(x.clone().into_diff())),
+                (Some(_), None) => match changes.last_mut() {
+                    Some(InnerVec::Remove { ref mut count }) => *count += 1,
+                    _ => changes.push(InnerVec::Remove { count: 1 }),
+                },
+            }
+        }
+
+        DiffVec(changes)
+    }
+
+    /// Merges a `DiffVec<T>`; `diff` with `self`; a `Vec<T>` to create a new `Vec<T>`.
+    fn merge(&self, diff: Self::Type) -> Self {
+        let mut vec: Self = self.clone();
+
+        for change in diff.into_iter() {
+            match change {
+                InnerVec::Add(d) => vec.push(<T>::from_diff(d)),
+                InnerVec::Change { index, item } => vec[index] = self[index].merge(item),
+                InnerVec::Remove { count } => {
+                    for _ in 0..count {
+                        vec.pop().expect("Unable to pop from the vector");
+                    }
+                }
+            }
+        }
+
+        vec
+    }
+
+    /// Converts a `DiffVec<T>`; `diff` into a `Vec<T>`.
+    fn from_diff(diff: Self::Type) -> Self {
+        let mut vec: Vec<T> = vec![];
+
+        for (_idx, elm) in diff.0.into_iter().enumerate() {
+            match elm {
+                InnerVec::Add(add) => vec.push(<T>::from_diff(add)),
+                InnerVec::Change { index: _, item } => {
+                    vec.push(<T>::from_diff(item));
+                }
+                _ => {}
+            }
+        }
+
+        vec
+    }
+
+    /// Converts a `Vec<T>` into a `DiffVec<T>`
+    fn into_diff(self) -> Self::Type {
+        let mut changes: Vec<InnerVec<T>> = vec![];
+        for inner in self {
+            changes.push(InnerVec::Add(inner.into_diff()));
+        }
+        DiffVec(changes)
+    }
+}
+
+/// Debug trait for `DiffVec<T>`
+impl<T: Diff> Debug for DiffVec<T> {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        write!(f, "DiffVec ")?;
+        f.debug_list().entries(self.0.iter()).finish()
+    }
+}
+
+/// Debug trait for `InnerVec<T>`
+impl<T: Diff> Debug for InnerVec<T> {
+    fn fmt(&self, f: &mut Formatter) -> FmtResult {
+        match &self {
+            Self::Change { index, item } => f
+                .debug_struct("Change")
+                .field("index", index)
+                .field("item", item)
+                .finish(),
+            Self::Remove { count } => f.debug_struct("Remove").field("count", count).finish(),
+            Self::Add(diff) => write!(f, "Add({:#?})", diff),
```

Comment:
Is there a `f.debug_` fn that could do this? Just to preserve formatting options